beacon-chain/blockchain/core.go

```diff
@@ -131,6 +146,46 @@ func (b *BeaconChain) CanProcessBlock(fetcher powchain.POWBlockFetcher, block *t
 	return validTime, nil
 }
 
+// RotateValidatorSet is called  every dynasty transition. It's primary function is
+// to go through queued validators and induct them to be active, and remove bad
+// active validator whose balance is below threshold to the exit set. It also cross checks
+// every validator's switch dynasty before induct or remove.
+func (b *BeaconChain) RotateValidatorSet() ([]types.ValidatorRecord, []types.ValidatorRecord, []types.ValidatorRecord) {
+
+	var newExitedValidator []types.ValidatorRecord
+	var newActiveValidators []types.ValidatorRecord
+
+	// Loop through active validator set, remove validator whose balance is below 50% and switch dynasty > current dynasty.
+	for _, validator := range b.state.CrystallizedState.ActiveValidators {
+
+		if validator.Balance < params.DefaultBalance/2 {
+			newExitedValidator = append(newExitedValidator, validator)
+		} else if validator.SwitchDynasty == b.CrystallizedState().Dynasty+1 {
+			newExitedValidator = append(newExitedValidator, validator)
+		} else {
+			newActiveValidators = append(newActiveValidators, validator)
+		}
+	}
+
+	// Get the total number of validator we can induct.
+	inductNum := b.ActiveValidatorCount()
```

Comment:
the spec says `up to a maximum of (len(crystallized_state.active_validators) // 30) + 1 `

should this be `b.ActiveValidatorCount() / 30 + 1`?